Improve editor selection while executing vault commnad

zmyrgel opened the pull request:
Imrove editor selection by first looking command pointed by VISUAL, then EDITOR
and falling back on using 'vi' if commands pointed b yVISUAL and EDITOR are not
found.
Also give better error message when no executable editor is not found in path.

##### SUMMARY
Current code doesn't look editor set by VISUAL env variable. VISUAL should take precedence over EDITOR. Also the code doesn't check if the actual command exists and is executable resulting in vague "File not found" error if no editor is found.
This fixes both issues and Fixes #34295

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ansible

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /home/tmy/.ansible.cfg
  configured module search path = [u'/home/tmy/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib64/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 22 2017, 11:31:09) [GCC 7.2.0]

```

##### ADDITIONAL INFORMATION

Commit:
improve editor selection

Imrove editor selection by first looking command pointed by VISUAL, then EDITOR
and falling back on using 'vi' if commands pointed b yVISUAL and EDITOR are not
found.
Also give better error message when no executable editor is not found in path.

Commit:
fix finding editor outside PATH

Editor given in EDITOR or VISUAL might not be present in PATH.
Add code to check if direct path to editor works first, then check if it can be
found in PATH.

Akasurde commented:
Test failing - https://app.shippable.com/github/ansible/ansible/runs/48659/2/console

zmyrgel commented:
If I'm reading that console output correctly its trying to mock some edit actions and can't find editor to make the changes. 
I guess its running in docker image, does the image have vi installed?

The question really is should the ansible check that the editor given in VISUAL/EDITOR is actually found and executable? But I'd like it at least give a better error message than misleading:
"ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory"

mattclay commented:
CI failure in the `vault` integration test:

```
2018-01-02 21:42:47 ERROR! failed to find an editor
2018-01-02 21:42:47 Could not find or access './faux-editor.py'
```

Commit:
less invasive editor command setting

look editor command from VISUAL, then EDITOR and fallback on 'vi'.
When running editor, catch OSError and report missing editor explicitly.

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/49034/5/tests

Commit:
restore accidentally removed 'raise' in _edit_file_helper

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -853,6 +853,7 @@ def _edit_file_helper(self, filename, secret,
         except:
```
actually, this should be a 'finally' instead (also remove 848)

zmyrgel replied:
I tried it with finally before but that seems to be wrong. If its finally, the tmp_path is removed after try-catch and the read_data call right after it fails.

bcoca replied:
we just might need to change the scope of where the 'finally' is set

zmyrgel replied:
pushed more stuff to try block and added finally clause. Seems better.
Also returned the mock object and fixed it to use "touch" as editor. Test seems to pass now but I think there should be still tests to see that VISUAL and EDITOR env variables are properly handled.

zmyrgel commented:
It think the code itself should be ok but I didn't look at the tests. Currently failing because the _editor_shell_command doesn't take filename arg anymore. I'm looking into fixing the test.

Commit:
use finally clause to shred tmp_path in _edit_file_helper

Commit:
no need to mock _editor_shell_command method in VaultEditor tests

zmyrgel commented:
Now the test fail with "no such editor 'vi'" message which seems to be the case.
I inspected the docker image and there is no vi editor installed.
Should the docker image be modified so that vi editor is present or should the test set VISUAL or EDITOR to some existing editor?

mattclay commented:
Since the test failures are in unit tests, the tests should be using mocks to present whatever editor is needed for each test scenario.

bcoca commented:
or just take the error as the 'correct' result from the test, as that is what i would expect if no vi is installed

Commit:
restore mocking of _editor_shell_command

make the mock call match the new _editor_shell_command method.

sivel commented:
This is something we would be willing to accept, however this PR needs to be rebased as there are merge conflicts.

needs_info

sivel commented:
Closing in favor of https://github.com/ansible/ansible/pull/51991
